Code/src/main/java/nl/utwente/viskell/ui/components/Block.java

```diff
@@ -116,51 +118,65 @@ public final CustomUIPane getPane() {
     }
 
     /**
+     * @return All InputAnchors of the block.
+     */
+    public List<InputAnchor> getAllInputs() {
+        return ImmutableList.of();
+    }
+
+    /**
+     * @return Only the active (as specified with by the knot index) inputs.
+     */
+    public List<InputAnchor> getActiveInputs() {
+        return getAllInputs();
+    }
+    
+    /**
+     * @return the optional output Anchor for this Block
+     */
+    public Optional<OutputAnchor> getOutputAnchor() {
+        return Optional.empty();
+    }
+    
+    /**
      * @return The expression this Block represents.
      * 
      * If the expression is not up-to-date it gets updated.
      */
-    public Expression getExpr() {
+    public final Expression getExpr() {
         // Assure expr is up-to-date.
         if (getExprIsDirty()) {
             updateExpr();
+            setExprIsDirty(false);
```

Comment:
Seems to be one other place, but cleaning up the whole getter/setter/property/visibility mess is a later step.